File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/api/PolarTrainingSessionApi.kt
package com.polar.sdk.api

import com.polar.sdk.api.model.PolarExerciseSession
import com.polar.sdk.api.model.trainingsession.PolarTrainingSession
import com.polar.sdk.api.model.trainingsession.PolarTrainingSessionFetchResult
import com.polar.sdk.api.model.trainingsession.PolarTrainingSessionReference
import kotlinx.coroutines.flow.Flow
import java.time.LocalDate

/**
 * Polar training session API.
 *
 * Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_TRAINING_DATA]
 */
interface PolarTrainingSessionApi {

    /**
     * Get training session references for a given period. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_TRAINING_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @param fromDate The starting date of the period to retrieve training session references from. Optional.
     * @param toDate The ending date of the period to retrieve training session references from. Optional.
     * @return A [Flow] emitting [PolarTrainingSessionReference] objects representing the training session references for the specified period.
     */
    fun getTrainingSessionReferences(
        identifier: String,
        fromDate: LocalDate? = null,
        toDate: LocalDate? = null
    ): Flow<PolarTrainingSessionReference>

    /**
     * Api for removing single training session from a Polar device. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_TRAINING_DATA]
     * You can get a list of training sessions with [getTrainingSessionReferences] API.
     *
     * @param identifier The Polar device ID or BT address.
     * @param reference PolarTrainingSessionReference with path in device to the training session to be removed.
     * @throws Throwable if the operation fails
     */
    suspend fun deleteTrainingSession(identifier: String, reference: PolarTrainingSessionReference)

    /**
     * Get training session. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_TRAINING_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @param trainingSessionReference The reference to the training session to retrieve.
     * @return [PolarTrainingSession] object representing the training session data.
     * @throws Throwable if the operation fails
     */
    suspend fun getTrainingSession(
        identifier: String,
        trainingSessionReference: PolarTrainingSessionReference
    ): PolarTrainingSession

    /**
     * Get training session with progress tracking. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_TRAINING_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @param trainingSessionReference The reference to the training session to retrieve.
     * @return A [Flow] emitting [PolarTrainingSessionFetchResult] objects with progress updates and final result.
     */
    fun getTrainingSessionWithProgress(
        identifier: String,
        trainingSessionReference: PolarTrainingSessionReference
    ): Flow<PolarTrainingSessionFetchResult>

    /**
     * Start an exercise session on the device. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_TRAINING_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @param profile The sport profile to use for the exercise session.
     * @throws Throwable if the operation fails
     */
    suspend fun startExercise(identifier: String, profile: PolarExerciseSession.SportProfile)

    /**
     * Pause an ongoing exercise session. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_TRAINING_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @throws Throwable if the operation fails
     */
    suspend fun pauseExercise(identifier: String)

    /**
     * Resume a paused exercise session. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_TRAINING_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @throws Throwable if the operation fails
     */
    suspend fun resumeExercise(identifier: String)

    /**
     * Stop the current exercise session. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_TRAINING_DATA]
     *
     * By default, the session is saved on the device.
     *
     * @param identifier The Polar device ID or BT address.
     * @throws Throwable if the operation fails
     */
    suspend fun stopExercise(identifier: String)

    /**
     * Get the current exercise session status from the device. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_TRAINING_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @return The current [PolarExerciseSession.ExerciseInfo] for the device.
     * @throws Throwable if the operation fails
     */
    suspend fun getExerciseStatus(identifier: String): PolarExerciseSession.ExerciseInfo

    /**
     * Observe exercise session status changes from the device. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_TRAINING_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @return A [Flow] emitting [PolarExerciseSession.ExerciseInfo] whenever the session status changes.
     * Errors may include [PolarErrors].
     */
    fun observeExerciseStatus(identifier: String): Flow<PolarExerciseSession.ExerciseInfo>
}

File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/api/PolarTrainingSessionApi.swift
///  Copyright © 2025 Polar. All rights reserved.

import Foundation

/// Protocol defining methods to access training session data.
public protocol PolarTrainingSessionApi {
    /// Get training session references for a given period. If fromDate and toDate are not given this
    /// method will return all training session references in the device.
    /// Otherwise returns list of training session references from the specified closed or
    /// half-open date interval.
    ///
    /// - Parameters:
    ///   - identifier: The Polar device ID or BT address.
    ///   - fromDate: The starting date of the period to retrieve training session references from. Optional.
    ///   - toDate: The ending date of the period to retrieve training session references from. Optional.
    /// - Requires SDK feature(s): `PolarBleSdkFeature.feature_polar_training_data`
    /// - Returns: An array of `PolarTrainingSessionReference` objects for the specified period.
    /// - Throws: See `PolarErrors` for possible errors.
    func getTrainingSessionReferences(
        identifier: String,
        fromDate: Date?,
        toDate: Date?
    ) async throws -> [PolarTrainingSessionReference]

    /// Get a specific training session using a reference.
    ///
    /// - Parameters:
    ///   - identifier: The Polar device ID or BT address.
    ///   - trainingSessionReference: The reference to the training session to retrieve.
    /// - Requires SDK feature(s): `PolarBleSdkFeature.feature_polar_training_data`
    /// - Returns: `PolarTrainingSession` containing the session data.
    /// - Throws: See `PolarErrors` for possible errors.
    func getTrainingSession(
        identifier: String,
        trainingSessionReference: PolarTrainingSessionReference
    ) async throws -> PolarTrainingSession

    /// Get a specific training session using a reference with progress updates.
    ///
    /// - Parameters:
    ///   - identifier: The Polar device ID or BT address.
    ///   - trainingSessionReference: The reference to the training session to retrieve.
    ///   - progressHandler: A closure called with `PolarTrainingSessionProgress` as the session data is being fetched.
    /// - Requires SDK feature(s): `PolarBleSdkFeature.feature_polar_training_data`
    /// - Returns: `PolarTrainingSession` containing the complete session data.
    /// - Throws: See `PolarErrors` for possible errors.
    func getTrainingSessionWithProgress(
        identifier: String,
        trainingSessionReference: PolarTrainingSessionReference,
        progressHandler: @escaping (PolarTrainingSessionProgress) -> Void
    ) async throws -> PolarTrainingSession

    /// Api for removing single training session from a Polar device. You can get a list of training sessions with `getTrainingSessionReferences` API.
    ///
    /// - Parameters:
    ///   - identifier: The Polar device ID or BT address.
    ///   - reference: PolarTrainingSessionReference with path in device to the training session to be removed.
    /// - Requires SDK feature(s): `PolarBleSdkFeature.feature_polar_training_data`
    /// - Throws: See `PolarErrors` for possible errors invoked.
    func deleteTrainingSession(identifier: String, reference: PolarTrainingSessionReference) async throws

    /// Start an exercise session on the device.
    ///
    /// - Parameters:
    ///   - identifier: The Polar device ID or BT address.
    ///   - profile: The sport profile to use for the exercise session.
    /// - Requires SDK feature(s): `PolarBleSdkFeature.feature_polar_training_data`
    /// - Throws: See `PolarErrors` for possible errors invoked.
    func startExercise(identifier: String, profile: PolarExerciseSession.SportProfile) async throws

    /// Pause an ongoing exercise session.
    ///
    /// - Parameter identifier: The Polar device ID or BT address.
    /// - Requires SDK feature(s): `PolarBleSdkFeature.feature_polar_training_data`
    /// - Throws: See `PolarErrors` for possible errors invoked.
    func pauseExercise(identifier: String) async throws

    /// Resume a paused exercise session.
    ///
    /// - Parameter identifier: The Polar device ID or BT address.
    /// - Requires SDK feature(s): `PolarBleSdkFeature.feature_polar_training_data`
    /// - Throws: See `PolarErrors` for possible errors invoked.
    func resumeExercise(identifier: String) async throws

    /// Stop the current exercise session.
    ///
    /// By default, the session is saved on the device.
    ///
    /// - Parameter identifier: The Polar device ID or BT address.
    /// - Requires SDK feature(s): `PolarBleSdkFeature.feature_polar_training_data`
    /// - Throws: See `PolarErrors` for possible errors invoked.
    func stopExercise(identifier: String) async throws

    /// Get the current exercise session status from the device.
    ///
    /// - Parameter identifier: The Polar device ID or BT address.
    /// - Requires SDK feature(s): `PolarBleSdkFeature.feature_polar_training_data`
    /// - Returns: The current `PolarExerciseSession.ExerciseInfo` for the device.
    /// - Throws: See `PolarErrors` for possible errors invoked.
    func getExerciseStatus(identifier: String) async throws -> PolarExerciseSession.ExerciseInfo

    /// Observe exercise session status notifications from the device.
    ///
    /// - Parameter identifier: The Polar device ID or BT address.
    /// - Requires SDK feature(s): `PolarBleSdkFeature.feature_polar_training_data`
    /// - Returns: `AsyncThrowingStream` emitting `PolarExerciseSession.ExerciseInfo` whenever the session status changes.
    func observeExerciseStatus(identifier: String) -> AsyncThrowingStream<PolarExerciseSession.ExerciseInfo, Error>
}
